rfcs/remote-schema-permissions.md

```diff
@@ -0,0 +1,244 @@
+# Remote schema permissions
+
+Currently, all remote schemas are fully exposed to all roles. We want to enable role-based access control to remote schemas as well.
+
+There are 2 things we want to focus on (there are possibly more):
+
+1. Role-based schema masking
+2. Role-based argument presets
+
+## Role-based schema masking
+
+A PoC of this feature was implemented in https://github.com/hasura/graphql-engine/pull/2690
+
+Schema masking is the ability to hide parts of the schema from the user. By default, we want to enforce "deny all" semantics by not exposing any type for a new role. 
+This role can then be allowed access to parts of the schema by specifying the types and fields it should have access to.
+
+The API for role-based schema proposed in #2690 looked like the following (**NOTE:** This is for illustration, not the final API as it doesn't incorporate argument presets):
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields: [hello, user]
+    - name: User
+      fields: [a,b,c]
+```
+
+You may notice that we will need to start with the root types `Query`, `Mutation`, `Subscription` (names might be different for other schemas) to give meaningful access to the role. In the above example, we are essentially saying that: only query fields `hello` (assume scalar type) and `user` (type `User`) are exposted to a role called "user". Further, the type `User` is also restricted to expose the fields `a, b, c` only.
+
+Another way of looking at role-based schema masking is essentially that we are creating role-based types. 
+
+### Behaviour for each type
+
+Schema masking can be done for any GraphQL type. We will describe the effect of schema masking on each type below:
+
+**Object**
+
+```
+allowed_objects:
+- name: Query
+  fields: [hello, user]
+- name: User
+  fields: [a,b,c]
+```
+
+Object types can specify the subset of fields in the type that are allowed in the schema.
+
+**Scalars**
+
+```
+allowed_custom_scalars:
+- name: UUID
+- name: Geography
+```
+
+Only `UUID` and `Geography` custom scalars are allowed in the schema. Note that built-in scalars like `Int`, `String`, etc cannot be masked. 
+
+**Interfaces**
+
+```
+allowed_interfaces:
+- name: Shape
+  fields: [height, width]
+```
+
+Interface type can also specify a subset of fields (similar to object types)
+
+**Unions**
+
+```
+allowed_unions:
+- name: SearchResult
+  objects: [Person]
+```
+
+Union types can specify the subset of list of object types.
+
+
+**Enums**
+
+```
+allowed_enums:
+- name: Direction
+  values: [North, South]
+```
+
+Enum types can specify a subset of values for the enum.
+
```

Comment:
Would these just replace any offending value with a `null` should they be returned? And/or will we filter nulls where we have a list, e.g. `[Direction]`, and so these would be like row-level permissions?

Reply:
Hmm, not sure. Should we worry about response validation as well?